File: solutions/python/linked-list/reverse-linked-list.py
# LeetCode 206 — Reverse Linked List (Easy)
# Category: Linked List · Approach: Iterative
# Time: O(n) | Space: O(1)
# Source: https://leetcode.com/problems/reverse-linked-list/

class ListNode:
    def __init__(self, val=0, nxt=None):
        self.val = val
        self.next = nxt

def reverse_list(head: ListNode | None) -> ListNode | None:
    prev = None
    while head:
        nxt = head.next
        head.next = prev
        prev = head
        head = nxt
    return prev

File: solutions/typescript/linked-list/reverse-linked-list.ts
// LeetCode 206 — Reverse Linked List (Easy)
// Category: Linked List · Approach: Iterative
// Time: O(n) | Space: O(1)
// Source: https://leetcode.com/problems/reverse-linked-list/

class ListNode {
  val: number;
  next: ListNode | null = null;
  constructor(val = 0, next: ListNode | null = null) {
    this.val = val;
    this.next = next;
  }
}

function reverseList(head: ListNode | null): ListNode | null {
  let prev: ListNode | null = null;
  while (head) {
    const next = head.next;
    head.next = prev;
    prev = head;
    head = next;
  }
  return prev;
}
